core/domain/stats_services.py

```diff
@@ -1024,3 +1026,237 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
+    saves it.
+
+    Args:
+        state_reference: str. This is used to refer to a state
+            in an exploration or question. For an exploration the
+            value will be equal to 'exp_id.state_name' and for question
+            this will be equal to 'question_id'.
+        entity_type: str. The type of entity i.e ENTITY_TYPE_EXPLORATION
+            or ENTITY_TYPE_QUESTION, which are declared in feconf.py.
+        learner_answer_info_id: str. The unique ID of the learner answer info
+            which needs to be deleted.
+"""
+    learner_answer_details = get_learner_answer_details(
+        state_reference, entity_type)
+    if not learner_answer_details:
+        raise utils.InvalidInputException(
+            'No learner answer details found with the given state '
+            'reference and entity')
+    learner_answer_details.delete_learner_answer_info(
+        learner_answer_info_id)
+    save_learner_answer_details(
+        learner_answer_details, state_reference, entity_type)
+
+
+def update_state_reference(
+        new_state_reference, old_state_reference, entity_type):
+    """Updates the state reference field of the model instance if the
+    and then saves it.
+
+    Args:
+        new_state_reference: str. The new state reference which needs to be
+            updated..
+        old_state_reference: str. The old state reference which needs to be
+            changed.
+        entity_type: str. The type of entity i.e ENTITY_TYPE_EXPLORATION
+            or ENTITY_TYPE_QUESTION, which are declared in feconf.py.
+    """
+    learner_answer_details = get_learner_answer_details(
+        old_state_reference, entity_type)
+    if not learner_answer_details:
+        raise utils.InvalidInputException(
+            'No learner answer details found with the given state '
+            'reference and entity')
+    learner_answer_details.update_state_reference(new_state_reference)
+    save_learner_answer_details(
+        learner_answer_details, old_state_reference, entity_type)
+
+
+def delete_learner_answer_details_for_exploration_state(
+        exp_id, state_name):
+    """Delete the LearnerAnswerDetailsModel instance of entity type
+    'exploration'provided state_reference, which is generated by the
+    given exp_id and state_name.
+
+    Args:
+        exp_id: str. The ID of the exploration.
+        state_name: str. The name of the state.
+    """
+    state_reference = (
+        stats_models.LearnerAnswerDetailsModel.get_state_reference_for_exploration( #pylint: disable=line-too-long
+            exp_id, state_name))
+    learner_answer_details_model = (
+        stats_models.LearnerAnswerDetailsModel.get_model_instance(
+            feconf.ENTITY_TYPE_EXPLORATION, state_reference))
+    if learner_answer_details_model:
+        learner_answer_details_model.delete()
+
+
+def delete_learner_answer_details_for_question_state(question_id):
+    """Delete the LearnerAnswerDetailsModel instance of entity type
```

Comment:
Ditto: Deletes the LearnerAnswerDetailsModel for the given question ID.

Reply:
Done